api/opentrons/drivers/smoothie_drivers/v3_0_0/serial_communication.py

```
@@ -0,0 +1,96 @@
+import serial
+from serial.tools import list_ports
+import contextlib
+
+DRIVER_ACK = b'ok\r\nok\r\n'
+RECOVERY_TIMEOUT = 10
+DEFAULT_SERIAL_TIMEOUT = 5
+SMOOTHIE_BOARD_NAME = 'FT232R'
+
+ERROR_KEYWORD = b'error'
+ALARM_KEYWORD = b'ALARM'
+
+BAUDRATE = 14400
+
+
+def get_ports(device_name):
+    '''Returns all serial devices with a given name'''
+    filtered_devices = filter(
+        lambda device: device_name in device[1],
+        list_ports.comports()
+    )
+    device_ports =  [device[0] for device in filtered_devices]
+    return device_ports
+
+
+@contextlib.contextmanager
+def serial_with_temp_timeout(serial_connection, timeout):
+    '''Implements a temporary timeout for a serial connection'''
+    saved_timeout = serial_connection.timeout
+    if timeout is not None:
+        serial_connection.timeout = timeout
+    yield serial_connection
+    serial_connection.timeout = saved_timeout
+
+
+def _parse_smoothie_response(response):
+    if ERROR_KEYWORD in response or ALARM_KEYWORD in response:
+        print("[SMOOTHIE ISSUE]: ", response)
+
+    if DRIVER_ACK in response:
+        parsed_response = response.split(DRIVER_ACK)[0]
+        return parsed_response
+    else:
+        return None
+
+def clear_buffer(serial_connection):
+    serial_connection.reset_input_buffer()
+
+def _write_to_device_and_return(cmd, device_connection):
+    '''Writes to a serial device.
+    - Formats command
+    - Wait for ack return
+    - return parsed response'''
+    command = cmd + '\r\n'
+    # print("writing to smoothie: {}".format(command.encode()))
+    device_connection.write(command.encode())
+    response = device_connection.read_until(DRIVER_ACK)
+    print('RESPONSE: ', response)
+
+    clean_response = _parse_smoothie_response(response)
+    return clean_response.decode()
+
+
+def _connect(port_name, baudrate):
+    return serial.Serial(port_name, baudrate=baudrate, timeout=DEFAULT_SERIAL_TIMEOUT)
+
+
+def _attempt_command_recovery(command, serial_conn):
```

Why do we recover instead of failing early?